### PR TITLE
feat: Adds Pluggable Auth support to ADC 

### PR DESCRIPTION
**Pull request by lsirac**
Adds the ability to retrieve 3rd party tokens through user defined executables. 

See go/pluggable-auth-design.

The implementation for the handler will come in a separate PR.

### CODE REVIEW

**Review by Neenu1995 — APPROVED** [review]

**Review by TimurSadykov — CHANGES_REQUESTED** [review]
Hey, 

Went through the code, without tests for now. Will do another pass later. 
You will see many nits on grammar articles, sorry about that :) Please apply for the rest of the code. Basically, owners before me had a higher bar for docs, I'm trying to match them. My english could backfire for sure, so let me know if i'm incorrect somewhere.

File: oauth2_http/java/com/google/auth/oauth2/ExecutableHandler.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    /** @return The full command (absolute path) used to retrieve 3rd party tokens. */
+    String getExecutableCommand();
+
+    /**
+     * @return The set of process-local environment variable mappings to be set for the script to
+     *     execute.
+     */
+    Map<String, String> getEnvironmentMap();
+
+    /** @return The timeout for waiting for the executable to finish, in milliseconds. */
```
[the remark this answers is not in the document]
Comment (lsirac) [reply oauth2_http/java/com/google/auth/oauth2/ExecutableHandler.java:53]: I think "The timeout..." makes more sense here

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 2]
```diff
@@ -413,7 +436,7 @@ protected AccessToken exchangeExternalCredentialForAccessToken(
             tokenUrl, stsTokenExchangeRequest, transportFactory.create().createRequestFactory());
 
     // If this credential was initialized with a Workforce configuration then the
-    // workforcePoolUserProject must passed to STS via the the internal options param.
+    // workforcePoolUserProject must be passed to STS via the internal options param.
```
[the remark this answers is not in the document]
Comment (lsirac) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java]: This is fixed with the merge.

File: oauth2_http/java/com/google/auth/oauth2/PluggableAuthCredentials.java [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+  /** Internal constructor. See {@link Builder}. */
+  PluggableAuthCredentials(Builder builder) {
+    super(builder);
+    this.config = (PluggableAuthCredentialSource) builder.credentialSource;
+
+    if (builder.handler != null) {
+      handler = builder.handler;
+    } else {
+      // TODO(lsirac): Initialize handler.
```
[the remark this answers is not in the document]
Comment (lsirac) [reply oauth2_http/java/com/google/auth/oauth2/PluggableAuthCredentials.java:203]: Yes. Will come in the next PR.

File: oauth2_http/java/com/google/auth/oauth2/PluggableAuthCredentials.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+ * GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE, GOOGLE_EXTERNAL_ACCOUNT_IMPERSONATED_EMAIL, and
+ * GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE.
+ *
+ * <p>Please see this repositories README for a complete executable request/response specification.
+ * </pre>
+ */
+public class PluggableAuthCredentials extends ExternalAccountCredentials {
+
+  /**
+   * Encapsulates the credential source portion of the configuration for PluggableAuthCredentials.
```
[the remark this answers is not in the document]
Comment (lsirac) [reply oauth2_http/java/com/google/auth/oauth2/PluggableAuthCredentials.java:96]: See above comment

File: oauth2_http/java/com/google/auth/oauth2/PluggableAuthCredentials.java [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+   *       "timeout_millis": 5000,
+   *       "output_file": "/path/to/generated/cached/credentials"
+   *     }
+   *   }
+   * }
+   * </pre>
+   */
+  static class PluggableAuthCredentialSource extends CredentialSource {
+
+    // The default timeout for waiting for the executable to finish (30 seconds).
```
[the remark this answers is not in the document]
Comment (lsirac) [reply oauth2_http/java/com/google/auth/oauth2/PluggableAuthCredentials.java:117]: It is "The executable" as it is a definite article

**Pull request merged**